joblib/func_inspect.py

```diff
@@ -180,19 +204,22 @@ def filter_args(func, ignore_lst, args=(), kwargs=dict()):
     args = list(args)
     if isinstance(ignore_lst, _basestring):
         # Catch a common mistake
-        raise ValueError('ignore_lst must be a list of parameters to ignore '
+        raise ValueError(
+            'ignore_lst must be a list of parameters to ignore '
             '%s (type %s) was given' % (ignore_lst, type(ignore_lst)))
     # Special case for functools.partial objects
     if (not inspect.ismethod(func) and not inspect.isfunction(func)):
         if ignore_lst:
             warnings.warn('Cannot inspect object %s, ignore list will '
-                'not work.' % func, stacklevel=2)
+                          'not work.' % func, stacklevel=2)
         return {'*': args, '**': kwargs}
-    arg_spec = inspect.getargspec(func)
-    arg_names = arg_spec.args
-    arg_defaults = arg_spec.defaults or {}
-    arg_keywords = arg_spec.keywords
+    arg_spec = getfullargspec(func)
+    arg_names = arg_spec.args + arg_spec.kwonlyargs
```

Comment:
At the moment I am not doing any checks to see whether the keyword-only arguments were indeed passed as keywords. Maybe I should since there is some checks for the number of arguments in this function already.

Note that there would be an error when the function is called but that may not happen if the function is already cached.

```python

import os
import shutil
from joblib import Memory

def func_with_kwonly_args(a, b, *, kw1='kw1', kw2='kw2'):
    return a, b, kw1, kw2

cachedir = '/tmp/joblib'
if os.path.exists(cachedir):
    shutil.rmtree(cachedir)

mem = Memory(cachedir=cachedir, verbose=0)
func_cached = mem.cache(func_with_kwonly_args)

# func_cached(1, 2, 3, kw2=4) would raise an exception
func_cached(1, 2, kw1=3, kw2=4)
func_cached(1, 2, 3, kw2=4)  # no exception anymore
```